### ISSUE TITLE
Add comparison tensor_ops

### ISSUE THREAD

**Issue by coreylowman** [coreylowman/dfdx#306]
- [ ] eq
- [ ] ne
- [ ] gt
- [ ] gte
- [ ] lt
- [ ] lte

functions should look like
```rust
fn eq<S: Shape, E: Dtype, D: Device<E>, T>(a: Tensor<S, E, D, T>, b: Tensor<S, E, D, T>) -> Tensor<S, bool, D, T>
```

**Comment by coreylowman** [comment]
Note that you won't be able to backprop through these - so they will always return a tensor with a NoneTape. This is the same as pytorch behavior here

**Comment by coreylowman** [comment]
There should also be a way for comparing against scalar values, but that can come later

**Comment by ViliamVadocz** [comment]
I assume this will also require `Device` with `Unit`, not `Dtype`. 

Currently code like this does not compile because of the trait bound.
```rust
fn eq<S: Shape, E: Unit, D: Device<E>, T>(
    a: Tensor<S, E, D, T>,
    b: Tensor<S, E, D, T>,
) -> Tensor<S, bool, D, T> {
  todo!()
}
```
```
8 | fn eq<S: Shape, E: Unit, D: Device<E>, T>(
  |                             ^^^^^^^^^ the trait `shape::Dtype` is not implemented for `E`
  ```

**Comment by coreylowman** [comment]
Ah yeah my mistake in the description. You'll need to use DeviceStorage instead of Device. (Device is just DeviceStorage + all the traits for tensor ops, but will require `E: Dtype`). We can't remove that restriction because pretty much every other tensor op requires Dtype for gradients.

So the idea would be generally:

```rust
// step 1: add an EqKernel trait
// NOTE: E here is Unit instead of Dtype
trait EqKernel<E: Unit>: DeviceStorage {
     fn forward(...)
     // NOTE: won't need a backward method since these won't support backprop
}

// step 2: add eq method to Tensor
impl<S: Shape, E: Unit, D: EqKernel<E>, T: Tape<D>> Tensor<S, E, D, T> {
    pub fn eq(self, other: Self) -> Tensor<S, bool, D, NoneTape> {
         todo!("call kernel forward methods on device");
    }
}

// step 3: add eq function that calls tensor method. This is mainly for discovery for functions & docstrings
pub fn eq<S: Shape, E: Unit, D: EqKernel<E>, T: Tape<D>>(a: Tensor<S, E, D, T>, b: Tensor<S, E, D, T>) -> Tensor<S, bool, E, NoneTape> {
    t.eq(b)
}

// in cpu_kernel.rs
impl<E: Unit> EqKernel for Cpu { ... }

// in cuda_kernel.rs
impl<E: Unit> EqKernel for Cuda {...}
```

Does that help?

**Comment by ViliamVadocz** [comment]
Is there a reason the only integer type which implements `Unit` is `usize`?
